src/rgen/Microsoft.Macios.Generator/DataModel/DelegateParameter.cs

```diff
@@ -0,0 +1,173 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.Macios.Generator.Extensions;
+
+namespace Microsoft.Macios.Generator.DataModel;
+
+/// <summary>
+/// Readonly structure that describes a parameter in a delegate. This class contains less information
+/// than Parameter since some of the extra fields make no sense in delegates.
+/// </summary>
+readonly struct DelegateParameter : IEquatable<DelegateParameter> {
+
+	/// <summary>
+	/// Parameter position in the method.
+	/// </summary>
+	public int Position { get; }
+
+	/// <summary>
+	/// Type of the parameter.
+	/// </summary>
+	public string Type { get; }
+
+	/// <summary>
+	/// Parameter name
+	/// </summary>
+	public string Name { get; }
+
+	/// <summary>
+	/// True if the parameter is optional
+	/// </summary>
+	public bool IsOptional { get; init; }
+
+	/// <summary>
+	/// True if a parameter is using the 'params' modifier.
+	/// </summary>
+	public bool IsParams { get; init; }
+
+	/// <summary>
+	/// True if the parameter represents the 'this' pointer.
+	/// </summary>
+	public bool IsThis { get; init; }
+
+	/// <summary>
+	/// True if the parameter is nullable.:w
+	/// </summary>
+	public bool IsNullable { get; init; }
+
+	/// <summary>
+	/// True if the parameter type is blittable.
+	/// </summary>
+	public bool IsBlittable { get; }
+
+	/// <summary>
+	/// Returns if the parameter type is a smart enum.
+	/// </summary>
+	public bool IsSmartEnum { get; init; }
+
+	/// <summary>
+	/// Returns if the parameter is an array type.
+	/// </summary>
+	public bool IsArray { get; init; }
+
+	/// <summary>
+	/// The reference type used.
+	/// </summary>
+	public ReferenceKind ReferenceKind { get; init; }
+
+	public DelegateParameter (int position, string type, string name, bool isBlittable)
+	{
+		Position = position;
+		Type = type;
+		Name = name;
+		IsBlittable = isBlittable;
+	}
+
+	public static bool TryCreate (IParameterSymbol symbol,
+		[NotNullWhen (true)] out DelegateParameter? parameter)
+	{
+		var type = symbol.Type is IArrayTypeSymbol arrayTypeSymbol
+			? arrayTypeSymbol.ElementType.ToDisplayString ()
+			: symbol.Type.ToDisplayString ().Trim ('?', '[', ']');
+
+		parameter = new (symbol.Ordinal, type, symbol.Name, symbol.Type.IsBlittable ()) {
+			IsOptional = symbol.IsOptional,
+			IsParams = symbol.IsParams,
+			IsThis = symbol.IsThis,
+			IsNullable = symbol.NullableAnnotation == NullableAnnotation.Annotated,
+			IsSmartEnum = symbol.Type.IsSmartEnum (),
+			IsArray = symbol.Type is IArrayTypeSymbol,
+			ReferenceKind = symbol.RefKind.ToReferenceKind (),
+		};
+		return true;
+	}
+
+	/// <inheritdoc/>
+	public bool Equals (DelegateParameter other)
```

Comment:
Would it make sense to use a `record` instead of `struct` - since I believe records get Equals, GetHashCode and ToString implementations automatically added by the compiler?

Reply:
The do, but they are ref types. We want to try and use structure to stay in the stack, we may need to later once everyworks review the memory allocation.